File: app/src/main/java/org/blitzortung/android/app/controller/HistoryController.java
package org.blitzortung.android.app.controller;

import android.app.Activity;
import android.provider.ContactsContract;
import android.view.View;
import android.widget.ImageButton;
import android.widget.Toast;

import com.annimon.stream.function.Consumer;

import org.blitzortung.android.app.AppService;
import org.blitzortung.android.app.R;
import org.blitzortung.android.data.DataChannel;
import org.blitzortung.android.data.DataHandler;
import org.blitzortung.android.data.provider.result.DataEvent;
import org.blitzortung.android.data.provider.result.ResultEvent;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

import lombok.val;


public class HistoryController {

    private final Collection<ImageButton> buttons;
    private ImageButton historyRewind;
    private ImageButton historyForward;
    private ImageButton goRealtime;
    private AppService appService;

    private ButtonColumnHandler buttonHandler;
    private DataHandler dataHandler;
    private Consumer<DataEvent> dataEventConsumer = event -> {
        if (event instanceof ResultEvent) {
            ResultEvent resultEvent = (ResultEvent) event;
            if (!resultEvent.hasFailed()) {
                setRealtimeData(resultEvent.containsRealtimeData());
            }
        }
    };

    public HistoryController(final Activity activity) {
        buttons = new ArrayList<>();

        setupHistoryRewindButton(activity);
        setupHistoryForwardButton(activity);
        setupGoRealtimeButton(activity);

        setRealtimeData(true);
    }

    public void setButtonHandler(ButtonColumnHandler buttonColumnHandler) {
        this.buttonHandler = buttonColumnHandler;
    }

    public void setRealtimeData(boolean realtimeData) {
        if (appService != null && dataHandler.isCapableOfHistoricalData()) {
            historyRewind.setVisibility(View.VISIBLE);
            int historyButtonsVisibility = realtimeData ? View.INVISIBLE : View.VISIBLE;
            historyForward.setVisibility(historyButtonsVisibility);
            goRealtime.setVisibility(historyButtonsVisibility);
        } else {
            historyRewind.setVisibility(View.INVISIBLE);
            historyForward.setVisibility(View.INVISIBLE);
            goRealtime.setVisibility(View.INVISIBLE);
        }
        updateButtonColumn();
    }

    private void setupHistoryRewindButton(final Activity activity) {
        historyRewind = (ImageButton) activity.findViewById(R.id.historyRew);
        buttons.add(historyRewind);
        historyRewind.setOnClickListener(v -> {
            if (dataHandler.rewInterval()) {
                disableButtonColumn();
                historyForward.setVisibility(View.VISIBLE);
                goRealtime.setVisibility(View.VISIBLE);
                updateButtonColumn();
                updateData();
            } else {
                Toast toast = Toast.makeText(activity.getBaseContext(), activity.getResources().getText(R.string.historic_timestep_limit_reached), Toast.LENGTH_SHORT);
                toast.show();
            }
        });
    }

    private void setupHistoryForwardButton(Activity activity) {
        historyForward = (ImageButton) activity.findViewById(R.id.historyFfwd);
        buttons.add(historyForward);
        historyForward.setVisibility(View.INVISIBLE);
        historyForward.setOnClickListener(v -> {
            if (dataHandler.ffwdInterval()) {
                if (dataHandler.isRealtime()) {
                    configureForRealtimeOperation();
                } else {
                    dataHandler.updateData();
                }
            }
        });
    }

    private void setupGoRealtimeButton(Activity activity) {
        goRealtime = (ImageButton) activity.findViewById(R.id.goRealtime);
        buttons.add(goRealtime);
        goRealtime.setVisibility(View.INVISIBLE);
        goRealtime.setOnClickListener(v -> {
            if (dataHandler.goRealtime()) {
                configureForRealtimeOperation();
            }
        });
    }

    private void configureForRealtimeOperation() {
        disableButtonColumn();
        historyForward.setVisibility(View.INVISIBLE);
        goRealtime.setVisibility(View.INVISIBLE);
        updateButtonColumn();
        appService.restart();
    }

    private void updateButtonColumn() {
        if (buttonHandler != null) {
            buttonHandler.updateButtonColumn();
        }
    }

    public Collection<ImageButton> getButtons() {
        return buttons;
    }

    private void disableButtonColumn() {
        buttonHandler.lockButtonColumn();
    }

    private void updateData() {
        dataHandler.updateData(Collections.singleton(DataChannel.STRIKES));
    }

    public void setAppService(AppService appService) {
        this.appService = appService;
        dataHandler = appService != null ? appService.getDataHandler() : null;
    }

    public Consumer<DataEvent> getDataConsumer() {
        return dataEventConsumer;
    }
}

File: app/src/main/java/org/blitzortung/android/app/controller/HistoryController.kt
/*

   Copyright 2015, 2016 Andreas Würl

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

*/

package org.blitzortung.android.app.controller

import android.content.Context
import android.view.View
import android.widget.ImageButton
import org.blitzortung.android.app.ButtonGroup
import org.blitzortung.android.app.databinding.MainBinding
import org.blitzortung.android.data.DataChannel
import org.blitzortung.android.data.MainDataHandler
import org.blitzortung.android.data.Mode
import org.blitzortung.android.data.provider.result.ResultEvent
import org.blitzortung.android.protocol.Event

class HistoryController(
    private val context: Context,
    private val binding: MainBinding,
    private val buttonHandler: ButtonColumnHandler<ImageButton, ButtonGroup>,
    private val dataHandler: MainDataHandler
) {

    private val buttons: MutableCollection<ImageButton> = arrayListOf()

    private var animationRunning = false

    val dataConsumer = { event: Event ->
        if (event is ResultEvent) {
            setRealtimeData(event.containsRealtimeData())
        }
    }

    init {
        setupStartStopAnimationButton()
        setupGoRealtimeButton()
        setRealtimeData(true)
    }

    private fun setRealtimeData(realtimeData: Boolean) {
        val historyButtonsVisibility = if (realtimeData || animationRunning) View.INVISIBLE else View.VISIBLE
        binding.goRealtime.visibility = historyButtonsVisibility
        if (realtimeData) {
            binding.timeSlider.progress = binding.timeSlider.max
        }
        updateButtonColumn()
    }

    private fun setupStartStopAnimationButton() {
        addButtonWithOnClickAction(binding.startStopAnimation) {
            if (animationRunning) {
                dataHandler.stop()
                binding.startStopAnimation.setImageResource(android.R.drawable.ic_media_play)
                dataHandler.goRealtime()
                configureForRealtimeOperation()

                animationRunning = false
                setRealtimeData(true)
                binding.timeSlider.isEnabled = true
            } else {
                dataHandler.stop()
                binding.startStopAnimation.setImageResource(android.R.drawable.ic_media_pause)
                animationRunning = true
                binding.timeSlider.isEnabled = false
                dataHandler.startAnimation()
            }
        }

        binding.startStopAnimation.visibility = View.VISIBLE
    }

    private fun setupGoRealtimeButton() {
        addButtonWithOnClickAction(binding.goRealtime) {
            if (dataHandler.goRealtime()) {
                configureForRealtimeOperation()
            }
        }
    }

    private fun addButtonWithOnClickAction(button: ImageButton, action: (View) -> Unit): ImageButton {
        return button.apply {
            buttons.add(this)
            visibility = View.INVISIBLE
            setOnClickListener(action)
        }
    }

    private fun configureForRealtimeOperation() {
        binding.goRealtime.visibility = View.INVISIBLE
        updateButtonColumn()

        dataHandler.restart()
        val historySteps = dataHandler.historySteps()
        binding.timeSlider.max = historySteps
        binding.timeSlider.progress = historySteps
    }

    private fun updateButtonColumn() {
        buttonHandler.updateButtonColumn()
    }

    fun getButtons(): Collection<ImageButton> {
        return buttons.toList()
    }

    private fun updateData() {
        dataHandler.updateData(setOf(DataChannel.STRIKES))
    }

    fun onResume() {
        animationRunning = dataHandler.mode == Mode.ANIMATION
        if (animationRunning) {
            binding.startStopAnimation.setImageResource(android.R.drawable.ic_menu_close_clear_cancel)
            binding.timeSlider.isEnabled = false
        }
    }
}
